lib/rack/builder.rb

```diff
@@ -31,6 +31,30 @@ module Rack
   # You can use +map+ to construct a Rack::URLMap in a convenient way.
 
   class Builder
+    # Config stores settings on what the server supports, such as whether it
+    # is multithreaded.
+    class Config
+      def initialize(multithread: true, reentrant: multithread)
+        @multithread = multithread
+        @reentrant = reentrant
+      end
+
+      # Whether the application server will invoke the application from multiple threads. Implies {reentrant?}.
+      def multithread?; @multithread; end
+
+      # Re-entrancy is a feature of event-driven servers which may perform non-blocking operations. When
+      # an operation blocks, that particular request may yield and another request may enter the application stack.
+      def reentrant?; @reentrant; end
```

Comment:
If `multithread?` is an implementation detail, so is `multifiber?`.  `parallel?` is not accurate on CRuby, since Ruby code does not execute in parallel on CRuby (though that is also an implementation detail).  Replacing `reentrant?` with `concurrent?` seems like a good idea to me, so I'll make that change.

I would like to get rid of `multithread?`, but I think it it is necessary until we drop Ruby 2.7 support. That is because on Ruby 2.7 and below, you cannot use Rack::Lock in the concurrent but not multithread case, since it uses Mutex internally.